File: 09_Count_substrings_with_exactly_K_distinct_characters.c
#include <stdio.h>
#include <string.h>

#define MAX_CHAR 26

int countAtMostK(char *s, int k) {
    int freq[MAX_CHAR] = {0};
    int left = 0, right = 0;
    int n = strlen(s);
    int count = 0;
    int unique = 0;

    while (right < n) {
        if (freq[s[right] - 'a'] == 0)
            unique++;
        freq[s[right] - 'a']++;

        while (unique > k) {
            freq[s[left] - 'a']--;
            if (freq[s[left] - 'a'] == 0)
                unique--;
            left++;
        }

        count += (right - left + 1);
        right++;
    }

    return count;
}

int countExactlyKDistinct(char *s, int k) {
    if (k == 0) return 0;
    return countAtMostK(s, k) - countAtMostK(s, k - 1);
}

int main() {
    char s[1000001];
    int k;
    
    scanf("%s", s);
    scanf("%d", &k);

    printf("%d\n", countExactlyKDistinct(s, k));

    return 0;
}

File: 09_Count_substrings_with_exactly_K_distinct_characters.py
def countAtMostK(s, k):
    from collections import defaultdict
    freq = defaultdict(int)
    left = 0
    count = 0
    for right in range(len(s)):
        freq[s[right]] += 1
        while len(freq) > k:
            freq[s[left]] -= 1
            if freq[s[left]] == 0:
                del freq[s[left]]
            left += 1
        count += right - left + 1
    return count

def countExactlyK(s, k):
    return countAtMostK(s, k) - countAtMostK(s, k - 1)

s = input()
k = int(input())
print(countExactlyK(s, k))